### PR TITLE
Feat/store encounter

### PR DESCRIPTION
**Pull request by JeffreyStocker**
adds the ability to store encounters in the bestiary page

### CODE REVIEW

File: bestiary.html [hunk 1]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+			<h3>Saved Encounters</h3>
+			<div class="mb-4">
+				<div class="row">
+					<ul class="list" id="listofsaves">
+						<!-- populate with JS -->
+					</ul>
+				</div>
+				<hr>
+				<div class="row">
+					<input class="ecgen__sv_newSaveName form-control col-8-4 input-xs" placeholder="New Encounter Name"></input>
```
Comment (TheGiddyLimit) [inline bestiary.html]: No camel case in class names; `ecgen__sv_newSaveName` -> e.g. `ecgen__sv-new-save-name`

File: bestiary.html [hunk 2]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+					<span>
+						<button class="btn btn-primary mr-2 btn-xs float-r" onclick="encounterBuilder.handleSaveClick(true)">New Save</button>
```
Comment (TheGiddyLimit) [inline bestiary.html]: Remove this `float-r` hack and use e.g. `col-2-6` on the outer `span` (and probably change that to a `div`)

File: bestiary.html [hunk 3]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+					</span>
+				</div>
+				<hr>
+				<div class="row">
+					<div class="col-12 flex" style="justify-content: center">
```
Comment (TheGiddyLimit) [inline bestiary.html]: Remove inline styling; use class `flex-vh-center` instead of `flex`

File: bestiary.html [hunk 4]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+						<button class="btn btn-primary mr-2 ecgen__sv_load" onclick="encounterBuilder.handleLoadClick()" disabled>Load</button>
```
Comment (TheGiddyLimit) [inline bestiary.html]: `ecgen__sv_load` -> `ecgen__sv-load`

File: bestiary.html [hunk 5]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+						<button class="btn btn-primary mr-2 ecgen__sv_save"  onclick="encounterBuilder.handleSaveClick()" disabled>Save</button>
```
Comment (TheGiddyLimit) [inline bestiary.html]: `ecgen__sv_save` -> `ecgen__sv-save`

File: bestiary.html [hunk 6]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+						<div style="width: 10px; height: 1px;"></div>
```
Comment (TheGiddyLimit) [inline bestiary.html]: Remove repeated inline styling; make this a class (e.g. `ecgen__sv-spacer`)

File: bestiary.html [hunk 7]
```diff
@@ -133,17 +133,47 @@ <h1 class="page__title">Bestiary</h1>
...
+						<button class="btn btn-danger mr-2" onclick="encounterBuilder.handleResetEncounterSavesClick()">Clear Saves</button>
+						<div style="width: 10px; height: 1px;"></div>
+						<button class="btn btn-primary mr-2 ecgen__sv_cancel">Cancel</button>
```
Comment (TheGiddyLimit) [inline bestiary.html]: `ecgen__sv_cancel` -> `ecgen__sv-cancel`

File: bestiary.html [hunk 8]
```diff
@@ -297,6 +327,9 @@ <h4 class="ecgen__raw_total">Total XP: ?</h4>
 						<button class="btn btn-primary btn-xs mr-2 ecgen__sv_file">Save to File</button>
 						<button class="btn btn-primary btn-xs mr-2 ecgen__ld_file">Load from File</button>
 						<div style="width: 10px; height: 1px;"></div>
+						<!-- <button class="btn btn-primary btn-xs mr-2 ecgen__sv_browser">Save</button> -->
```
Comment (TheGiddyLimit) [inline bestiary.html]: Remove commented line

File: bestiary.html [hunk 9]
```diff
@@ -297,6 +327,9 @@ <h4 class="ecgen__raw_total">Total XP: ?</h4>
...
+						<button class="btn btn-primary btn-xs mr-2 ecgen__ld_browser">Load/Save to Browser</button>
```
Comment (TheGiddyLimit) [inline bestiary.html]: `ecgen__ld_browser` -> `ecgen__ld-browser`

File: js/bestiary-encounterbuilder.js [hunk 10]
```diff
@@ -6,7 +6,10 @@ class EncounterBuilder {
 		this._cache = null;
 		this._lastPlayerCount = null;
 		this._advanced = false;
-
+		this._SAVEDENCOUNTERSAVELOCATION = "ENCOUNTER_SAVED_STORAGE";
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: This is unreadable, add some underscores; `this._SAVEDENCOUNTERSAVELOCATION` -> `this._SAVED_ENCOUNTER_SAVE_LOCATION`

File: js/bestiary-encounterbuilder.js [hunk 11]
```diff
@@ -66,6 +69,16 @@ class EncounterBuilder {
 			});
 		});
 		$(`.ecgen__reset`).click(() => confirm("Are you sure?") && encounterBuilder.pReset());
+
+		$('.ecgen__ld_browser').click(() => {
+			encounterBuilder.uiLoadMenuToggle(true);
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Single-line handler can be inlined

File: js/bestiary-encounterbuilder.js [hunk 12]
```diff
@@ -66,6 +69,16 @@ class EncounterBuilder {
...
+		});
+
+		$('.ecgen__sv_cancel').click(() => {
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Single-line handler can be inlined

File: js/bestiary-encounterbuilder.js [hunk 13]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
...
+		} else {
+			namesHtml = names.map((name, index) => {
+				let className = "";
+				if (name === this._savedName) {
+					className = "list-multi-selected";
+					hasAName = true;
+				}
+				return `<li class="${className}" onclick="encounterBuilder.selectSaveName(this, '${names[index]}')">
+					<span class="name">${name}</span>
+					<span style="float: right" class="no-wrap" onclick="event.preventDefault()">
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Avoid using float

File: js/bestiary-encounterbuilder.js [hunk 14]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
 			</span>
 		`;
 	}
+
+	async getSavedEncounters () {
+		let savedEncounters = await StorageUtil.pGet(this._SAVEDENCOUNTERSAVELOCATION);
+		if (savedEncounters === null || savedEncounters === undefined) {
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: simplify to `== null`

File: js/bestiary-encounterbuilder.js [hunk 15]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
 			</span>
 		`;
 	}
+
+	async getSavedEncounters () {
+		let savedEncounters = await StorageUtil.pGet(this._SAVEDENCOUNTERSAVELOCATION);
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Use `const` over `let`

File: js/bestiary-encounterbuilder.js [hunk 16]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
...
+			$('.ecgen__sv_load').prop('disabled', isDisabled).text('Reload');
+		} else {
+			$('.ecgen__sv_save').prop('disabled', isDisabled).text('Save');
+			$('.ecgen__sv_load').prop('disabled', isDisabled).text('Load');
+		}
+	}
+
+	uiLoadMenuGenerator () {
+		let namesHtml;
+		let names = Object.keys(this._savedEncounters);
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Use `const` over `let`

File: js/bestiary-encounterbuilder.js [hunk 17]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
...
+		} else if (encounterBuilder._savedName === encounterBuilder._selectedSave) {
+			name = this._savedName;
+		} else if (confirm("are you sure? You will overwrite this save file")) {
+			name = this._savedName;
+		} else {
+			return;
+		}
+
+		this._savedName = name;
+		let saveableState = this.getSaveableState();
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Use `const` over `let`

File: js/bestiary-encounterbuilder.js [hunk 18]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
...
+	async handleLoadClick () {
+		if (window.confirm("Are you sure?\n You will overwrite the current encounter!")) {
+			this.pDoLoadState(this._selectedSavedEncounter);
+			this.uiLoadMenuToggle(false);
+		}
+	}
+
+	handleDeleteClick (node) {
+		if (confirm(`Delete ${this._selectedSave}?`)) {
+			let name = $(node).parents("li").children('.name').text();
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Use `const` over `let`

File: js/bestiary-encounterbuilder.js [hunk 19]
```diff
@@ -933,6 +950,136 @@ class EncounterBuilder {
...
+			delete this._savedEncounters[name];
+			if (name === this._savedName) this._savedName = null;
+			this.setSavedEncounters();
+			this.uiLoadMenuGenerator();
+		}
+	}
+
+	handleResetEncounterSavesClick () {
+		if (confirm("The will clear ALL saved encounter! Are you sure?") && confirm("Are you very sure you want to delete all saved encounters?")) {
```
Comment (TheGiddyLimit) [inline js/bestiary-encounterbuilder.js]: Typo `saved encounter!` -> `saved encounters`

Remove double-prompt

File: scss/style.scss [hunk 20]
```diff
@@ -1234,6 +1234,10 @@ footer {
 	float: none;
 }
 
+.float-r {
```
Comment (TheGiddyLimit) [inline scss/style.scss]: Remove hacky float class

**Comment by JeffreyStocker** [comment]
All of the code has been fixed.

There are now some inconsistencies with the class naming scheme. The ones you have are written as "ecgen__sv_file" where mine are now labeled with '-' ie 'ecgen__sv-save'. 

I put all my logic into EnounterBuilder. But if you want to use the saves in other places (ie DM screen) then we may want to pull some of it out and put it into the utils.js file.

**Comment by TheGiddyLimit** [comment]
Lovely, thanks! I'll check it out tomorrow. The naming thing, yeah, I fucked up and used underscores a bunch instead of dashes. It's supposed to be BEM, though.

**Comment by TheGiddyLimit** [comment]
Good enough. I've reworked it and integrated it into the main repo; it'll be bundled into the next update. Closing to avoid the inevitable double-merge.

**Pull request closed without merging**